nixos/modules/services/web-apps/peertube.nix

```diff
@@ -0,0 +1,431 @@
+
+      script = let
+        psqlSetupCommands = pkgs.writeText "peertube-init.sql" ''
+          SELECT 'CREATE USER "${cfg.database.user}"' WHERE NOT EXISTS (SELECT FROM pg_roles WHERE rolname = '${cfg.database.user}')\gexec
+          SELECT 'CREATE DATABASE "${cfg.database.name}" OWNER "${cfg.database.user}" TEMPLATE template0 ENCODING UTF8' WHERE NOT EXISTS (SELECT FROM pg_database WHERE datname = '${cfg.database.name}')\gexec
+          \c '${cfg.database.name}'
+          CREATE EXTENSION IF NOT EXISTS pg_trgm;
+          CREATE EXTENSION IF NOT EXISTS unaccent;
+        '';
+      in "${config.services.postgresql.package}/bin/psql -f ${psqlSetupCommands}";
+
+      serviceConfig = {
+        Type = "oneshot";
+        WorkingDirectory = cfg.package;
+        # User and group
+        User = "postgres";
+        Group = "postgres";
+        # Sandboxing
+        RestrictAddressFamilies = [ "AF_UNIX" ];
+        MemoryDenyWriteExecute = true;
+        # System Call Filtering
+        SystemCallFilter = "~@clock @cpu-emulation @debug @keyring @memlock @module @mount @obsolete @privileged @raw-io @reboot @resources @setuid @swap";
+      } // cfgService;
+    };
+
+    systemd.services.peertube = {
+      description = "PeerTube daemon";
+      after = [ "network.target" ]
+        ++ lib.optionals cfg.redis.createLocally [ "redis.service" ]
+        ++ lib.optionals cfg.database.createLocally [ "postgresql.service" "peertube-init-db.service" ];
+      wantedBy = [ "multi-user.target" ];
+
+      environment.NODE_CONFIG_DIR = "/var/lib/peertube/config";
+      environment.NODE_ENV = "production";
+      environment.NODE_EXTRA_CA_CERTS = "/etc/ssl/certs/ca-certificates.crt";
+      environment.HOME = cfg.package;
+
+      path = with pkgs; [ bashInteractive ffmpeg nodejs openssl yarn youtube-dl ];
+
+      serviceConfig = {
+        Type = "simple";
+        ExecStartPre = let preStartScript = pkgs.writeScript "peertube-pre-start.sh" ''
+          #!/bin/sh
+          umask 077
+          cat > /var/lib/peertube/config/local.yaml <<EOF
+          ${lib.optionalString ((!cfg.database.createLocally) && (cfg.database.passwordFile != null)) ''
+          database:
+            password: '$(cat ${cfg.database.passwordFile})'
+          ''}
+          ${lib.optionalString (cfg.redis.passwordFile != null) ''
+          redis:
+            auth: '$(cat ${cfg.redis.passwordFile})'
+          ''}
+          ${lib.optionalString (cfg.smtp.passwordFile != null) ''
+          smtp:
+            password: '$(cat ${cfg.smtp.passwordFile})'
+          ''}
+          EOF
+          ln -sf ${cfg.package}/config/default.yaml /var/lib/peertube/config/default.yaml
+          ln -sf ${configFile} /var/lib/peertube/config/production.json
+        '';
+        in "${preStartScript}";
+        ExecStart = let startScript = pkgs.writeScript "peertube-start.sh" ''
+          #!/bin/sh
+          exec npm start
+        '';
+        in "${startScript}";
```

Comment:
Why don't we just call `npm start` instead of wrapping it like this?

Reply:
Fixed, changed to - `ExecStart = "${pkgs.nodejs-14_x}/bin/npm start";`